packages/netlify-cms-core/src/reducers/collections.js

```diff
@@ -97,8 +97,19 @@ const selectors = {
   },
 };
 
-export const selectFields = (collection, slug) =>
-  selectors[collection.get('type')].fields(collection, slug);
+export const selectFields = (collection, slug, entry = null) => {
+  if (entry === null) {
+    return selectors[collection.get('type')].fields(collection, slug);
+  }
+  const fields = selectors[collection.get('type')].fields(collection, slug);
+  let entryFields = fields;
+  entry.get('data', []).forEach((value, key) => {
+    entryFields = !fields.find(field => field.get('name') === key)
+      ? entryFields.push(Map({ name: key, widget: 'hidden' }))
+      : entryFields;
+  });
+  return entryFields;
+};
```

Comment:
Rather than conditionally altering the behavior of `selectFields`, we can operate more explicitly by creating a separate `selectNonConfiguredFields` selector. Here's a modified version that avoids mutable practices like the `let` keyword (note: just wrote it up, didn't test):

```js
export const selectNonConfiguredFields = (collection, entry) => {
  const fields = selectFields(collection, entry.slug);
  return entry.get('data', [])
    .filter(entryField => fields.every(field => entryField.get('name') !== field.get('name')))
    .map((field, key) => Map({ name: key, widget: 'hidden' }))
};
```

Reply:
I was going back and forth on splitting this guy out, originally. I'm going to make 2 selectors: ```selectNonConfiguredFields``` which only returns the fields that are different between the entry and the collection and ```selectAllFields``` which will do the concatenation.